client/asset/btc/spv_peer_manager.go

```diff
@@ -0,0 +1,308 @@
+}
+
+// PeerManagerChainService are the functions needed for an SPVPeerManager
+// to communicate with a chain service.
+type PeerManagerChainService interface {
+	AddPeer(addr string) error
+	RemoveNodeByAddr(addr string) error
+	Peers() []SPVPeer
+}
+
+// SPVPeerManager implements peer management functionality for all bitcoin
+// clone SPV wallets.
+type SPVPeerManager struct {
+	cs PeerManagerChainService
+
+	peersMtx sync.RWMutex
+	peers    map[string]*walletPeer
+
+	savedPeersFilePath string
+
+	defaultPeers []string
+
+	defaultPort string
+
+	log dex.Logger
+}
+
+func (w *SPVPeerManager) connectedPeers() map[string]interface{} {
+	peers := w.cs.Peers()
+	connectedPeers := make(map[string]interface{})
+	for _, peer := range peers {
+		connectedPeers[peer.Addr()] = struct{}{}
+	}
+	return connectedPeers
+}
+
+// Peers returns the list of peers that the wallet is connected to. It also
+// returns the peers that the user added that the wallet may not currently
+// be connected to.
+func (w *SPVPeerManager) Peers() ([]*asset.WalletPeer, error) {
+	w.peersMtx.RLock()
+	defer w.peersMtx.RUnlock()
+
+	connectedPeers := w.connectedPeers()
+
+	walletPeers := make([]*asset.WalletPeer, 0, len(connectedPeers))
+
+	for originalAddr, peer := range w.peers {
+		_, connected := connectedPeers[peer.resolvedName]
+		delete(connectedPeers, peer.resolvedName)
+		walletPeers = append(walletPeers, &asset.WalletPeer{
+			Addr:      originalAddr,
+			Connected: connected,
+			Source:    peer.source.toAssetPeerSource(),
+		})
+	}
+
+	for peer := range connectedPeers {
+		walletPeers = append(walletPeers, &asset.WalletPeer{
+			Addr:      peer,
+			Connected: true,
+			Source:    asset.Discovered,
+		})
+	}
+
+	return walletPeers, nil
+}
+
+// resolveAddress resolves an address to ip:port. This is needed because neutrino
+// internally resolves the address, and when neutrino is called to return its list
+// of peers, it will return the resolved addresses. Therefore, we call neutrino
+// with the resolved address, then we keep track of the mapping of address to
+// resolved address in order to be able to display the address the user provided
+// back to the user.
+func (w *SPVPeerManager) resolveAddress(addr string) (string, error) {
+	host, strPort, err := net.SplitHostPort(addr)
+	if err != nil {
+		switch err.(type) {
+		case *net.AddrError:
+			host = addr
+			strPort = w.defaultPort
+		default:
+			return "", err
+		}
+	}
+
+	// Tor addresses cannot be resolved to an IP, so just return onionAddr
+	// instead.
+	if strings.HasSuffix(host, ".onion") {
+		return host, nil
+	}
+
+	// Attempt to look up an IP address associated with the parsed host.
+	ips, err := net.LookupIP(host)
+	if err != nil {
+		return "", err
+	}
+
+	if len(ips) == 0 {
+		return "", fmt.Errorf("no addresses found for %s", host)
+	}
+
+	port, err := strconv.Atoi(strPort)
+	if err != nil {
+		return "", err
+	}
+
+	return (&net.TCPAddr{
+		IP:   ips[0],
+		Port: port,
+	}).String(), nil
```

Comment:
Without `net.TCPAddr.Zone` set, this is just `net.JoinHostPort(IP, Itoa(Port))`.  Since we've already done `SplitHostPort` at the top, it seems like we have a needless Atoi -> Itoa in the mix.

`return net.JoinHostPort(ips[0], strPort), nil`?